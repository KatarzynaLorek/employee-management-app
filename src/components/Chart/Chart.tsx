import React from 'react';
import { Bar, Doughnut } from 'react-chartjs-2';
import { ChartType, DepartmentType, PositionType, SortType } from '../../types/charts';
import { IResponseObject } from '../../types/responses';

interface IChartProps {
  sortType: SortType;
  chartType: ChartType;
  data: IResponseObject[];
}

const Chart = ({ sortType, chartType, data }: IChartProps): JSX.Element => {
  const departments: DepartmentType[] = ['accountancy', 'IT', 'sales', 'marketing'];
  const positions: PositionType[] = ['junior', 'regular', 'senior', 'manager'];

  //Filter function
  const employeesCountByKey = (
    sortType: 'department' | 'departmentOnly' | 'position' | 'positionOnly',
    firstSortKey: DepartmentType | PositionType,
    secondSortKey?: DepartmentType | PositionType,
  ): number => {
    switch (sortType) {
      case 'department':
        return data
          .filter((employee) => employee.department === firstSortKey)
          .filter((employee) => employee.position === secondSortKey).length;
      case 'departmentOnly':
        return data.filter((employee) => employee.department === firstSortKey).length;
      case 'position':
        return data
          .filter((employee) => employee.position === firstSortKey)
          .filter((employee) => employee.department === secondSortKey).length;
      case 'positionOnly':
        return data.filter((employee) => employee.position === firstSortKey).length;
    }
  };

  //Data getters
  const getEmployeesByDepartmentOnly = (): number[] => {
    return departments.map((department) => employeesCountByKey('departmentOnly', department));
  };

  const getEmployeesByDepartment = (department: DepartmentType): number[] => {
    return positions.map((position) => employeesCountByKey('position', position, department));
  };

  const getEmployeesByPositionOnly = (): number[] => {
    return positions.map((position) => employeesCountByKey('positionOnly', position));
  };

  const getEmployeesByPosition = (position: PositionType): number[] => {
    return departments.map((department) => employeesCountByKey('department', department, position));
  };

  //Chart data objects
  const BarChartData = data && {
    labels:
      sortType === 'department'
        ? ['Accountancy', 'IT', 'Marketing', 'Sales']
        : ['Junior', 'Regular', 'Senior', 'Manager'],
    datasets: [
      {
        label: sortType === 'department' ? 'Junior' : 'Accountancy',
        data:
          sortType === 'department'
            ? getEmployeesByPosition('junior')
            : getEmployeesByDepartment('accountancy'),
        backgroundColor: ['#003f5c'],
      },
      {
        label: sortType === 'department' ? 'Regular' : 'IT',
        data:
          sortType === 'department'
            ? getEmployeesByPosition('regular')
            : getEmployeesByDepartment('IT'),
        backgroundColor: ['#7a5195'],
        borderWidth: 0,
      },
      {
        label: sortType === 'department' ? 'Senior' : 'Marketing',
        data:
          sortType === 'department'
            ? getEmployeesByPosition('senior')
            : getEmployeesByDepartment('marketing'),
        backgroundColor: ['#ef5675'],
        borderWidth: 0,
      },
      {
        label: sortType === 'department' ? 'Manager' : 'Sales',
        data:
          sortType === 'department'
            ? getEmployeesByPosition('manager')
            : getEmployeesByDepartment('sales'),
        backgroundColor: ['#ffa600'],
        borderWidth: 0,
      },
    ],
  };

  const DoughnutChartData = data && {
    labels:
      sortType === 'department'
        ? ['Accountancy', 'IT', 'Marketing', 'Sales']
        : ['Junior', 'Regular', 'Senior', 'Manager'],
    datasets: [
      {
        data:
          sortType === 'department' ? getEmployeesByDepartmentOnly() : getEmployeesByPositionOnly(),
        backgroundColor: ['#003f5c', '#7a5195', '#ef5675', '#ffa600'],
      },
    ],
  };

  return chartType === 'bar' ? (
    <Bar type="bar" data={BarChartData} options={{ maintainAspectRatio: false }} />
  ) : (
    <Doughnut type="doughnut" data={DoughnutChartData} options={{ maintainAspectRatio: false }} />
  );
};
export default Chart;
