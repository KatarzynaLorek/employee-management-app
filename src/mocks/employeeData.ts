import { IResponseObject } from '../types/responses';

export const mockEmployeeData: IResponseObject = {
  id: '1',
  firstName: 'mockFirstName',
  lastName: 'mockLastname',
  position: 'junior',
  department: 'IT',
};
