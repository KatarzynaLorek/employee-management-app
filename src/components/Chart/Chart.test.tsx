import React from 'react';
import { render } from '@testing-library/react';
import Chart from './Chart';
import { mockEmployeeData } from '../../mocks/employeeData';

jest.mock('react-chartjs-2', () => ({
  Bar: () => null,
  Doughnut: () => null,
}));

describe('Chart component', () => {
  it('renders bar chart with sorting by department', () => {
    const container = render(
      <Chart sortType="department" chartType="bar" data={[mockEmployeeData]} />,
    );
    expect(container.baseElement).toMatchSnapshot();
  });
  it('renders bar chart with sorting by position', () => {
    const container = render(
      <Chart sortType="position" chartType="bar" data={[mockEmployeeData]} />,
    );
    expect(container.baseElement).toMatchSnapshot();
  });
  it('renders doughnut chart with sorting by department', () => {
    const container = render(
      <Chart sortType="department" chartType="doughnut" data={[mockEmployeeData]} />,
    );
    expect(container.baseElement).toMatchSnapshot();
  });
  it('renders doughnut chart with sorting by position', () => {
    const container = render(
      <Chart sortType="position" chartType="doughnut" data={[mockEmployeeData]} />,
    );
    expect(container.baseElement).toMatchSnapshot();
  });
});
