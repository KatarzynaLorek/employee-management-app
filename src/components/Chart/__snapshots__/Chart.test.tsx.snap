// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart component renders bar chart with sorting by department 1`] = `
<body>
  <div />
</body>
`;

exports[`Chart component renders bar chart with sorting by position 1`] = `
<body>
  <div />
</body>
`;

exports[`Chart component renders doughnut chart with sorting by department 1`] = `
<body>
  <div />
</body>
`;

exports[`Chart component renders doughnut chart with sorting by position 1`] = `
<body>
  <div />
</body>
`;
